login: add show/hide password toggle

Rameshkumar31595 opened the pull request:
## Related Issue
Fixes: #5083

## Summary
Implements a show/hide password toggle button on the JupyterHub login form.

## Why
At 2i2c, users mentioned difficulty identifying typos in passwords during login.  
This PR adds a small “Show” button next to the password field, improving usability and accessibility.

## Implementation
- Added a Bootstrap input group for password + button alignment.
- Button toggles input type between `password` and `text`.
- Includes accessible ARIA attributes:
  - `aria-pressed` reflects toggle state.
  - `aria-controls="password_input"` links button to field.
- Text label dynamically switches between “Show” and “Hide”.

## Accessibility
- Fully keyboard accessible.
- Screen reader friendly (`aria-label` and state updates).
- No visual or behavioral regressions.

## Testing
- Verified manually in browser.
- Local text-based test (`tests_local/test_login_toggle.py`) passed successfully.

## Notes
This change only affects the `share/jupyterhub/templates/login.html` file.

minrk commented on share/jupyterhub/templates/login.html:
```diff
@@ -50,14 +49,48 @@ <h1>Sign in</h1>
+                                 transform: translateY(-50%);
+                                 border: none;
+                                 background: none;
+                                 cursor: pointer"
```
cursor: pointer is not needed, it's already set for buttons. We only need this if there's _not_ a button element.

minrk commented on share/jupyterhub/templates/login.html:
```diff
@@ -105,11 +137,9 @@ <h1>Sign in</h1>
   {{ super() }}
   <script>
     if (!window.isSecureContext) {
-      // unhide http warning
```
more comments removed. Is there a reason?

Rameshkumar31595 commented on share/jupyterhub/templates/login.html:
```diff
@@ -50,14 +49,48 @@ <h1>Sign in</h1>
+                         class="form-control"
+                         autocomplete="current-password"
+                         name="password"
+                         aria-describedby="toggle_password_visibility"
```
Thanks for the clarification, @minrk!
That makes sense — using aria-label sounds more appropriate here. I’ll update the implementation to replace aria-describedby with a dynamic aria-label that reflects the current visibility state (“Password visible” / “Password hidden”) for better screen reader support.

Rameshkumar31595 commented:
Hi @minrk,

I’ve completed all the requested updates:

Restored previously removed template comments for clarity

Replaced aria-describedby with dynamic aria-label for proper accessibility

Removed unnecessary inline CSS (cursor: pointer, positioning)

Used Bootstrap’s input-group for a consistent and accessible layout

Verified keyboard navigation and screen reader behavior for password visibility

Maintained the original file structure and block hierarchy

All checks have passed except for two browser-based tests (pytest (3.11, subdomain, browser) and pytest (3.13, browser)), which appear to be unrelated to the template changes.

The PR has no conflicts and can be cleanly merged.

Please review the latest version when you have time and consider it for merging.

Thank you for the detailed review and guidance throughout this contribution.

Rameshkumar31595 commented:
Hi @minrk, 

I’ve completed all the requested updates:

Restored the {# Allow full override... #} comment to its correct position

Simplified the password visibility toggle to only update aria-pressed and title

Verified _xsrf and login_error are correctly positioned

 All checks have passed and the PR can be cleanly merged.

When you have a chance, could you please review the latest version and consider merging

Thanks again for your thoughtful reviews and guidance, really appreciate that

Rameshkumar31595 commented:
Hi @minrk, @manics  I’ve manually verified all requested fixes, restored the comment to its correct position, and confirmed all checks are passing with no conflicts.
Please review the latest changes and let me know if anything else is needed, or consider merging if everything looks good. Thanks!
